### ISSUE TITLE
Deduce category from expense title

### ISSUE THREAD

**Issue by mertd** [spliit-app/spliit#72]
Automating the category selection would allow users to input and categorize their expenses faster and more reliably. Imagine a user typing their expense title, and based on their input, a category gets selected for them.

Two approaches of implementing this could be:

1. AI: Use the OpenAI integration added in #69. Prompt the latest stable turbo model for one of the available categories.
2. Keywords: For each category, maintain a set of keywords. If a keyword is found, the mapped category will be selected.

Implementing the AI option would allow categorization of many more titles than the keyword approach. It would also increase the running cost of spliit. Starting out with an implementation of keyword based mapping will be more economical. However, usability may be limited considering the range of possible titles.

I would be happy to give either of these approaches a try, with AI being my preference. Are you interested in such a feature? If yes, which approach are you leaning towards?

**Comment by scastiel** [comment]
Cool idea! I’m okay with testing with OpenAI.

The cost should be pretty low for each expense, but might grow fast if we call OpenAI automatically (~50 to 100 expenses are created each day currently, and it’s growing).

What do you think of adding a subtle magic wand icon at the right of the category, and call OpenAI only when clicked? At least as a first step.

**Comment by justcallmelarry** [comment]
I would like to see this as an opt-in feature, at least for self hosters (like the current OpenAI integration for pics)

To make fewer calls to OpenAI both the solutions could be used, ie (pseudo code):
```
if (any keywords in title) {
    category = hit
} else {
    category = get_category_from_openai()
}
```

In the long run:
With growing data for spliit.app, I think this is something that you could have a "buddy service" for, where you can have a model trained on the database data (you have everyones expense title and what category they gave it, while excluding all "General" categories)

**Comment by scastiel** [comment]
Agree that the feature should be opt-in.

I suggest we start with a plain call to OpenAI. An optimization for later could be to persist the results to avoid doing the same call too many times. As you suggest @justcallmelarry, on the long term it could be interesting to train our own model.

**Comment by mertd** [comment]
Sounds good, I will give the AI option a go then!

> ~What do you think of adding a subtle magic wand icon at the right of the category, and call OpenAI only when clicked? At least as a first step.~ Edit: let’s do that only later if the cost is too high.

Waiting until the expense title input loses focus may be a good balance between number of API calls and necessary user interactions.

> With growing data for spliit.app, I think this is something that you could have a "buddy service" for, where you can have a model trained on the database data

We could add support for custom endpoints compatible with the OpenAI API in a future iteration. Similar to #71, it would be beneficial for self-hosting and to avoid vendor lock-in, be it for a different AI vendor or possibly a future custom model.

**Comment by scastiel** [comment]
> Waiting until the expense title input loses focus may be a good initial balance between number of API calls and necessary user interactions.

👍

**Comment by mertd** [comment]
I created a first attempt at #80. Looking forward to your feedback!

**Comment by scastiel** [comment]
Resolved by #80. If we need to improve the category deduction (using keywords for instance), we’ll create a new issue.
